source/common/http/filter/gzip_filter.cc

```diff
@@ -0,0 +1,186 @@
+  return memory_level_ > 0 ? memory_level_ : DEFAULT_MEMORY_LEVEL;
+}
+
+uint64_t GzipFilterConfig::minimumLength() const {
+  return content_length_ > 29 ? content_length_ : MINIMUM_CONTENT_LENGTH;
+}
+
+uint64_t GzipFilterConfig::windowBits() const {
+  return (window_bits_ > 0 ? window_bits_ : DEFAULT_WINDOW_BITS) | GZIP_HEADER_VALUE;
+}
+
+GzipFilter::GzipFilter(GzipFilterConfigSharedPtr config)
+    : skip_compression_{true}, compressed_data_(), compressor_(), config_(config) {}
+
+void GzipFilter::onDestroy() {}
+
+FilterHeadersStatus GzipFilter::decodeHeaders(HeaderMap& headers, bool) {
+  // TODO(gsagula): The current implementation checks for the presence of 'gzip' and if the same
+  // is followed by Qvalue. Since gzip is the only available encoding right now, order/priority of
+  // preferred server encodings is disregarded (RFC2616-14.3).
+  skip_compression_ =
+      !(headers.AcceptEncoding() &&
+        std::regex_search(headers.AcceptEncoding()->value().c_str(), acceptEncodingRegex()));
+  return FilterHeadersStatus::Continue;
+}
+
+FilterHeadersStatus GzipFilter::encodeHeaders(HeaderMap& headers, bool end_stream) {
+  if (!end_stream && !skip_compression_ && isMinimumContentLength(headers) &&
+      isContentTypeAllowed(headers) && isCacheControlAllowed(headers) && isEtagAllowed(headers) &&
+      isLastModifiedAllowed(headers) && isTransferEncodingAllowed(headers) &&
+      !headers.ContentEncoding()) {
+    headers.removeContentLength();
+    headers.insertContentEncoding().value(Http::Headers::get().ContentEncodingValues.Gzip);
+    compressor_.init(config_->compressionLevel(), config_->compressionStrategy(),
+                     config_->windowBits(), config_->memoryLevel());
+  } else {
+    skip_compression_ = true;
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+FilterDataStatus GzipFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (skip_compression_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  const uint64_t n_data{data.length()};
+
+  if (n_data) {
+    compressor_.compress(data, compressed_data_);
+  }
+
+  if (end_stream) {
+    compressor_.flush(compressed_data_);
+  }
+
+  if (compressed_data_.length()) {
+    data.drain(n_data);
+    data.move(compressed_data_);
+    return Http::FilterDataStatus::Continue;
+  }
+
+  return Http::FilterDataStatus::StopIterationNoBuffer;
+}
+
+bool GzipFilter::isMinimumContentLength(HeaderMap& headers) const {
+  if (headers.ContentLength()) {
+    uint64_t content_length{};
+    StringUtil::atoul(headers.ContentLength()->value().c_str(), content_length);
+    return content_length >= config_->minimumLength();
+  }
+
+  return headers.TransferEncoding() &&
+         headers.TransferEncoding()->value().find(
+             Http::Headers::get().TransferEncodingValues.Chunked.c_str());
+}
+
+bool GzipFilter::isContentTypeAllowed(HeaderMap& headers) const {
+  if (config_->contentTypeValues().size() && headers.ContentType()) {
```

Comment:
use .empty() to check whether it's empty, rather than taking the size(), which might in theory be a non-trivial calculation, depending on the STL impl.